### PR TITLE
FEATURE: TNT-1344 Download exported report/dash via ObjectURL

### PR DESCRIPTION
**Pull request by xMort**
Currently, exported report or dashboard task is polled via fetch API. When task is not ready, JSON body with polling uri is returned. When task finishes, export content is returned in response body. Fetch utility always parsed the response body to text.

Now, when export task is polled, response body is parsed to ArrayBuffer. When the export content is returned, buffer is converted to Blob and then ObjectURL is created for the Blob. ObjectURL creation attaches the blob to window instance and creates a URL that points to it. This URL is then used in anchor element appended to DOM. Script simulates the click on the anchor, file is downloaded to user's filesystem, and anchor is removed from DOM, and ObjectURL revoked to release blob data from window instance and memory.

This approach saves one additional call to backend when the content of export is fetched again. Most importantly, it works around the issues with authentication of this GET request made by click on the anchor in the case when token authentication is used (no cookies) or UserAgentReduction mode is enabled by the user's browser.

Name of the download must be parsed from the content-disposition header, otherwise API uri value would be used.

JIRA: TNT-1344

---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended check plugins`                                | Dashboard plugins tests                                    |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11628/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13683/ : FAILURE in 16m 57s

File: libs/api-model-bear/api/api-model-bear.api.md [hunk 1]
```diff
@@ -1044,6 +1044,10 @@ export namespace GdcExport {
     }
     // (undocumented)
     export interface IExportResponse {
+        // (undocumented)
+        fileName?: string;
+        // (undocumented)
+        objectUrl?: string;
```
Comment (ivanmjartan) [inline libs/api-model-bear/api/api-model-bear.api.md:1050]: What is difference between  objectUrl and uri?

Comment (xMort) [reply libs/api-model-bear/api/api-model-bear.api.md:1050]: `uri` points to our API endpoint from which you can fetch the result (again, as it was already fetched once when polling resolved). `objectUrl` points to blob of data attached to the current `window` instance. The blob contains data from the polling result. The Object URL looks like this: `blob:https://localhost:8443/51b3d300-65ba-41e0-9bd2-7a23a75c571b` for example. Also, it is necessary to revoke it when you do not needed to release the blob data from memory and `window` instance. 

This could be an issue of this solution as it will be created every time you create export and you must not forget to revoke it, otherwise you will gather memory. Not sure if I should not do another method for exporting with this Object URL? Possibly put somewhere some `try finaly` block to catch possible errors and revoke the URL if it would not come to the actual download? I'm not sure that could happen however. Once you parse response and create blob, there are not many places when the logic could fail IMHO.

File: libs/api-client-bear/src/util.ts [hunk 2]
```diff
@@ -35,6 +36,7 @@ export interface IPollingOptions {
     attempts?: number;
     maxAttempts?: number;
     pollStep?: number;
+    blobContentType?: string;
```
Comment (ivanmjartan) [inline libs/api-client-bear/src/util.ts:39]: Can you describe each property ? I know that it  internal type but it will be handy ...

File: libs/api-client-bear/src/utils/export.ts [hunk 3]
```diff
@@ -1,7 +1,15 @@
...
+    const taskState = response.status;
     return taskState === SUCCESS_REQUEST_STATUS || taskState >= BAD_REQUEST_STATUS; // OK || ERROR
 }
+
+export const parseFileNameFromContentDisposition = (response: Response): string | undefined => {
+    const contentDispositionHeader = response.headers.get("Content-Disposition") || "";
+    // eslint-disable-next-line regexp/no-unused-capturing-group
+    const matches = /filename\*?=([^']*'')?([^;]*)/.exec(contentDispositionHeader);
+    const urlEncodedFileName = matches ? matches[2] : undefined;
+    return urlEncodedFileName ? decodeURIComponent(urlEncodedFileName) : undefined;
```
Comment (ivanmjartan) [inline libs/api-client-bear/src/utils/export.ts:14]: Is undefined fallback correct here ? Should be there some generic name  like exportData or something ?

File: libs/sdk-backend-tiger/src/utils/downloadFile.ts [hunk 4]
```diff
@@ -10,3 +12,11 @@ export function downloadFile(fileName: string, data: any) {
     URL.revokeObjectURL(link.href);
     document.body.removeChild(link);
 }
+
+export const parseNameFromContentDisposition = (response: AxiosResponse) => {
```
Comment (ivanmjartan) [inline libs/sdk-backend-tiger/src/utils/downloadFile.ts:16]: This method is identical to method in [libs/api-client-bear/src/utils/export.ts]

File: libs/sdk-ui-dashboard/api/sdk-ui-dashboard.api.md [hunk 5]
```diff
@@ -1606,7 +1607,7 @@ export interface DashboardInsightWidgetExportResolved extends IDashboardEvent {
 
 // @beta
 export interface DashboardInsightWidgetExportResolvedPayload {
-    resultUri: string;
+    result: IExportResult;
```
Comment (ivanmjartan) [inline libs/sdk-ui-dashboard/api/sdk-ui-dashboard.api.md:1610]: describe this property

File: libs/api-client-bear/src/report/report.ts [hunk 6]
```diff
@@ -60,7 +60,13 @@ export class ReportModule {
             .post(`/gdc/internal/projects/${projectId}/exportResult`, { body: requestPayload })
             .then((response: ApiResponse) => response.getData())
             .then((data: GdcExport.IExportResponse) =>
-                handleHeadPolling(this.xhr.get.bind(this.xhr), data.uri, isExportFinished, pollingOptions),
+                handleHeadPolling(this.xhr.get.bind(this.xhr), data.uri, isExportFinished, {
+                    ...pollingOptions,
+                    blobContentType:
+                        exportConfig.format === "xlsx"
```
[the remark this answers is not in the document]
Comment (ivanmjartan) [reply libs/api-client-bear/src/report/report.ts:66]: What about other formats like PDF ? The blobContentType is the same ???

File: libs/sdk-backend-tiger/src/utils/downloadFile.ts (hunk 4 above)
Comment (xMort) [reply libs/sdk-backend-tiger/src/utils/downloadFile.ts:16]: I had it originally shared but it is a bit different. One is accepting `AxiosResponse` the second fetch's `Response`. I did not want to unpack header value outside of it as I would have to do it three times, if I remember correctly.

However, the actual parsing with regexes could be shared. What would be the correct place for such a util as I require to use it in both Tiger and Panther API clients?

File: libs/api-client-bear/src/report/report.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (xMort) [reply libs/api-client-bear/src/report/report.ts:66]: For PDF it is "application/pdf". I can hide it to method for sure.

File: libs/api-client-bear/src/utils/export.ts (hunk 3 above)
Comment (xMort) [reply libs/api-client-bear/src/utils/export.ts:14]: If `undefined` is used, it will fallback on URI value in the downloader util, but I can use something generic. However, it will be missing the extension. I would leave it as is as it is unlikely exporter would stop sending `content-disposition` header with file name.

File: libs/sdk-ui-dashboard/api/sdk-ui-dashboard.api.md (hunk 5 above)
Comment (xMort) [reply libs/sdk-ui-dashboard/api/sdk-ui-dashboard.api.md:1610]: I updated the comment as it was wrong but strange that it is not here in the exported interface doc as the prop was  already documented. See `libs/sdk-ui-dashboard/src/model/events/insight.ts:64`

File: libs/api-client-bear/src/util.ts (hunk 2 above)
Comment (xMort) [reply libs/api-client-bear/src/util.ts:43]: Fixed

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-backend-tiger/src/utils/downloadFile.ts (hunk 4 above)
Comment (xMort) [reply libs/sdk-backend-tiger/src/utils/downloadFile.ts:16]: On the other hand, possibly it made sense that regex parsing is duplicated as the response is served by two different service (export service for Bear and Tiger) and we could keep these separate if one of them would change. This is however highly theoretical. If there is place where it can be put to have it shared, I will gladly change it.

File: libs/sdk-ui-dashboard/src/model/events/insight.ts [hunk 7]
```diff
@@ -658,15 +659,15 @@ export interface DashboardInsightWidgetExportResolved extends IDashboardEvent {
...
+    result: IExportResult,
     correlationId?: string,
 ): DashboardInsightWidgetExportResolved {
     return {
         type: "GDC.DASH/EVT.INSIGHT_WIDGET.EXPORT_RESOLVED",
         ctx,
         correlationId,
         payload: {
-            resultUri,
+            result,
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/model/events/insight.ts:670]: Can this be changed if this is beta? Possibly some plugins would require update?
Maybe I should change the payload to:

````
payload: {
	resultUri,
	result,
},
````

This way it should be compatible. There's just the issue with `objectURL` that will no longer be valid once download finishes.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11691/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13733/ : FAILURE in 16m 01s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11693/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13735/ : SUCCESS in 16m 02s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11696/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13738/ : SUCCESS in 19m 51s

File: libs/api-client-bear/package.json [hunk 8]
```diff
@@ -94,6 +94,7 @@
         "ts-loader": "^8.3.0",
         "typescript": "4.0.2",
         "webpack": "^5.58.0",
-        "webpack-cli": "^4.9.2"
+        "webpack-cli": "^4.9.2",
+        "blob-polyfill": "^7.0.20220408"
```
Comment (kandl) [inline libs/api-client-bear/package.json:98]: For which browsers do wee need this?

File: rush.json [hunk 9]
```diff
@@ -212,7 +212,8 @@
          * your PR branch, and in this situation "rush change" will also automatically invoke "git fetch"
          * to retrieve the latest activity for the remote master branch.
          */
-        "url": "git@github.com:gooddata/gooddata-ui-sdk.git"
+        "url": "git@github.com:gooddata/gooddata-ui-sdk.git",
+        "defaultBranch": "master" // branch used when "rush change" is called (current rush default value is "master")
```
Comment (kandl) [inline rush.json:216]: Why is this needed when it's same as the default value?

File: libs/api-client-bear/package.json (hunk 8 above)
Comment (xMort) [reply libs/api-client-bear/package.json:98]: This is not require by browser but by jest suite. Some tests ends up calling the export functions where `Blob` is created. The test suite uses `node` environment but `Blob` is available only in `jsdom`.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: rush.json (hunk 9 above)
Comment (xMort) [reply rush.json]: Good catch. The current value is `main`. I updated the comment. I was rewriting it few times and screw it up in the end anyway 😅

**Review by kandl — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3505

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/366/ : FAILURE in 1m 51s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/3638/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/2482/ : SUCCESS in 11m 06s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13741/ : SUCCESS in 16m 40s
